### DISCUSSION TITLE
RuntimeError: Internal radio Status (2, 1) OpError 0x20 sx126x.py micropython/lora

### DISCUSSION THREAD

**Discussion by davefes** [micropython/micropython#13619] [Libraries & Drivers]
Does anybody know where I can find these error codes.  For further details see:
https://github.com/orgs/micropython/discussions/13305

**Comment by davefes** [comment]
OK, I have found a `0x20`  at line 216 in `sx126x.py`

It seems to suggest that the tcxo is not being setup correctly or maybe is non-existent.

Looking through all the data I can find:
https://www.cdebyte.com/products/E77-400M22S
the term `tcxo` does not appear.

`README.md` suggests that `dio3_tcxo_millivolts` is not required.

Has anyone configured a TCXO or suggest what I am missing in setting up my module correctly?

**Reply by gampam2000** [reply]
Hi,
did you find a solution for this error? I have a similar error:

```
  File "/lib/lora/async_modem/async_modem.py", line 65, in send
  File "/lib/lora/sx126x.py", line 616, in prepare_send
  File "/lib/lora/sx126x.py", line 332, in _check_error
RuntimeError: Internal radio Status (2, 1) OpError 0x20
```

my lora cfg looks like this:
```
    lora_cfg = {
        "freq_khz": 869525,
        "sf": 7,
        "bw": "250",  # kHz
        "coding_rate": 5,
        "preamble_len": 8,
        "output_power": 0,
        "implicit_header": False,
        "crc_en": True,
        "syncword" : 0x1424,
        "dio3_tcxo_millivolts" : 1800
     }
```

Thanks & Best regards

**Reply by davefes** [reply]
Good to hear there are other people trying to get this to work.  See:
https://github.com/orgs/micropython/discussions/13110
Which module are you working on?

**Reply by gampam2000** [reply]
Hi, i use a ESP32 with an SX1262 module from Ebyte.
It works perfectly fine with the micropython lora driver from: https://github.com/ehong-tl/micropySX126X

So I know that the SPI connection and Lora works perfectly fine.

But with the driver from micopython-lib I get this `RuntimeError: Internal radio Status (2, 1) OpError 0x20`

I wanted to switch to this driver because of the async interface.

**Reply by davefes** [reply]
I have used:
https://github.com/Wei1234c/SX127x_driver_for_MicroPython_on_ESP8266
successful on five LoRa systems, ESP32 and RFM96W.

I spent many hours trying to get `micropython-lora` to work on the E77-400M22S.  The `build-NUCLEO_WL55` has
`micropython-lora` baked-in and seems to run on the E77-400M22S until that error.

I am tempted to post on the forum at this site as he seems to do a lot of LoRa work.
https://mischianti.org/

**Comment by davefes** [comment]
@gampam2000, did you make any further progress?

**Reply by gampam2000** [reply]
Yes I fixed it:

dio3_tcxo_millivolts=1800 is a parameter for the constructor not for the lora_cfg!

see https://github.com/micropython/micropython-lib/issues/870 my latest comment, there i also included my code.

I hope it helps you too.

**Reply by davefes** [reply]
@gampam2000, unfortunately I can't modify the constructor unless I build it myself.  There is another issue with the STM32 NUCLEO_WL55, which makes debugging (and using this part) quite difficult.  
https://github.com/orgs/micropython/discussions/13160

**Reply by gampam2000** [reply]
You can't change the micropython code?

**Reply by davefes** [reply]
The main problem is that I can only upload a 400-500 byte file as there is a conflict in communications as Andrew suggests,

I did upload:
```py
from lora import WL55SubGhzModem

LORA_FREQ_KHZ = 433500

def get_modem():
    global LORA_FREQ_KHZ

    lora_cfg = { 'freq_khz'		: LORA_FREQ_KHZ,
                'dio3_tcxo_millivolts' : '1800',
               }
    return WL55SubGhzModem(lora_cfg)

modem = get_modem()

print('finished')
```
and got the error.  There might be an issue with how the manufacturer expects the internal oscillators are set up, but no information was given.

Really until there is a way to upload even small files 1-3KB there is not much point persisting.
